### PR TITLE
Enable multislice in launch script

### PR DESCRIPTION
**Pull request by blahBlahhhJ**
Adds `node_count` in config to launch jobs with multislice training.

Refactored the launch script to use `queued-resources` API.

mention #629

### CODE REVIEW

**Comment by blahBlahhhJ** [comment]
Some UX design discussions: what behavior do we want our launch script to have?
1. * Before: with retry>0, the launch script will retry even if the job finishes without error
    *  After: the launch script will exit (no retry) if the job finishes without error
2. * Before: (not using `queued-resources` API, not sure about the actual behavior)
    * After: If `autodelete` is True, **and the job finishes with no error,** the launch script will tear down the TPU automatically (so that it doesn't eat up the TPU quota). Do we want this to happen when the launch script finishes all the retries as well? e.g. tear down the TPU after the job errors and retries 100 times.
 3. If the user manually kills the script (e.g. by ctrl-c), the TPU will still stay in the `queued-resources` list, taking up some quota even after suspended. Is there something we can do here?

File: infra/launch.py [hunk 1]
```diff
@@ -84,16 +107,16 @@ def start_tpu_vm(tpu_name, *, tpu_type, capacity_type, version, zone, autodelete
...
         f"--accelerator-type={tpu_type}",
-        f"--version={version}",
-        "--zone=" + zone,
+        f"--runtime-version={version}",
+        f"--zone={zone}",
         "--quiet",
     ]
     if capacity_type == "preemptible":
-        command.append("--preemptible")
+        command.append("--best-effort")
```
Comment (dlwh) [inline infra/launch.py:119]: let's add best-effort as a synonym for preemptible in the argparse

File: infra/launch.py [hunk 2]
```diff
@@ -102,8 +125,33 @@ def start_tpu_vm(tpu_name, *, tpu_type, capacity_type, version, zone, autodelete
...
+        command.append(f"--node-id={tpu_name}")
+    else:
+        command.append(f"--node-count={node_count}")
+
     cli.run_command(*command)
 
+    # wait for queued resource to complete
+    waited = 0
+    while True:
+        time.sleep(60)
```
Comment (dlwh) [inline infra/launch.py:139]: before sleeping this long we should log that we're gonna wait.

File: infra/launch.py [hunk 3]
```diff
@@ -102,8 +125,33 @@ def start_tpu_vm(tpu_name, *, tpu_type, capacity_type, version, zone, autodelete
         pass
     else:
         raise ValueError(f"Unknown capacity type: {capacity_type}")
+
+    if node_count == 1:
+        command.append(f"--node-id={tpu_name}")
+    else:
+        command.append(f"--node-count={node_count}")
```
Comment (dlwh) [inline infra/launch.py:132]: set `node-prefix`  to tpu_name?

File: infra/helpers/cli.py [hunk 4]
```diff
@@ -12,6 +12,11 @@ def run_command(*args, **kwargs):
     return subprocess.check_call(args, **kwargs)
 
 
+def run_command_parallel(*args, **kwargs):
```
Comment (dlwh) [inline infra/helpers/cli.py:15]: not really parallel but background yeah?

**Review by dlwh — COMMENTED** [review]
let's wait on @rjpower (maybe timing out after a couple of days) but other than comments lgtm

**Comment by dlwh** [comment]
To answer your questions:

1. new behavior is better
2. `Use the existing argument autodelete` is better. Existing behavior is kinda weird imho
3. I think the current behavior is the right behavior, especially with TRC TPUs. Maybe we should add a cleanup script or something but I think it's fine.

File: infra/launch.py (hunk 3 above)
Comment (blahBlahhhJ) [reply infra/launch.py:132]: The tpu name will automatically become the node prefix.

File: infra/launch.py [hunk 5]
```diff
@@ -38,44 +40,65 @@ def setup_vm_docker(tpu_name, zone, docker_base_image):
...
-        [
-            "gcloud",
-            "alpha",
-            "compute",
-            "tpus",
-            "tpu-vm",
-            "list",
-            "--zone=" + zone,
-        ]
+    return json.loads(
```
Comment (rjpower) [inline infra/launch.py:43]: Does this also show non-queued-resources TPUs? (Or should we always use queued-resources?)

**Review by rjpower — APPROVED** [review]
Seems good to me, thanks for doing this!

It seems like this will only work with queued resources now, but @dlwh would know if there were any real downsides of that. Adjusting things to merge both types of TPUs seems tedious and probably not worth it, assuming queued-resources works fine for smaller slices.

File: infra/launch.py (hunk 3 above)
Comment (dlwh) [reply infra/launch.py:132]: but you're not appending the tpu name?

Comment (dlwh) [reply infra/launch.py:132]: oh i see because it's the queued resource name.

**Review by dlwh — APPROVED** [review]

File: infra/launch.py (hunk 5 above)
Comment (blahBlahhhJ) [reply infra/launch.py:43]: after the migration everything will be in queued resources so we can probably just clean up all the non queued resources tpus and move on.

**Comment by blahBlahhhJ** [comment]
(waiting for a preemption event to see if my script retries correctly and push a final commit addressing @dlwh 's comments before I merge)

**Comment by dlwh** [comment]
will just deleting the TPU (not the resource) simulate preemption well enough?

**Comment by dlwh** [comment]
@blahBlahhhJ is this ready to merge?

**Pull request merged**